src/feature/NodeGraphFeature.ts

```diff
@@ -0,0 +1,183 @@
+      };
+
+      if ((this.connectionManager.status !== ConnectionStatus.Running) && (this.connectionManager.status !== ConnectionStatus.Starting)) {
+        if (this.shownLanguageServerNotAvailable) {
+          vscode.window.showInformationMessage("The Puppet Node Graph Preview is not available as the Editor Service is not ready");
+          this.shownLanguageServerNotAvailable = true;
+        }
+        return "The Puppet Node Graph Preview is not available as the Editor Service is not ready";
+      }
+
+      // Use the language server to render the document
+      return this.connectionManager.languageClient
+        .sendRequest(CompileNodeGraphRequest.type, sourceUri)
+        .then(
+          (compileResult) => {
+
+          var svgContent = '';
+          if (compileResult.dotContent !== null) {
+            var styling = `
+            bgcolor = "transparent"
+            color = "white"
+            rankdir = "TB"
+            node [ shape="box" penwidth="2" color="#e0e0e0" style="rounded,filled" fontname="Courier New" fillcolor=black, fontcolor="white"]
+            edge [ style="bold" color="#f0f0f0" penwith="2" ]
+
+            label = ""`
+
+            var graphContent = compileResult.dotContent;
+            // vis.jz sees backslashes as escape characters, however they are not in the DOT language.  Instead
+            // we should escape any backslash coming from a valid DOT file in preparation to be rendered
+            graphContent = graphContent.replace(/\\/g,"\\\\");
+            graphContent = graphContent.replace(`label = "vscode"`,styling);
+
+            svgContent = viz(graphContent,"svg");
+          }
+
+          var errorContent = `<div style='font-size: 1.5em'>${compileResult.error}</div>`
+          if (compileResult.error === null) { errorContent = ''; }
+
+          if (reporter) {
+            reporter.sendTelemetryEvent(PuppetNodeGraphToTheSideCommandId);
+          }
+
+          return `
+            ${errorContent}
+            <div id="graphviz_svg_div">
+              ${svgContent}
+            </div>`;
+      });
+    });
+  }
+
+  get onDidChange(): vscode.Event<vscode.Uri> {
+    return this.onDidChangeEvent.event;
+  }
+
+  public update(uri: vscode.Uri) {
+    if (!this.waiting) {
+      this.waiting = true;
+      setTimeout(() => {
+        this.waiting = false;
+        this.onDidChangeEvent.fire(uri);
+      }, 300);
+    }
+  }
+}
+
+export class NodeGraphFeature implements IFeature {
+  private provider: NodeGraphContentProvider;
+
+  constructor(
+    langID: string,
+    connectionManager: IConnectionManager,
+    logger: ILogger,
+    context: vscode.ExtensionContext
+  ) {
+    context.subscriptions.push(vscode.commands.registerCommand(PuppetNodeGraphToTheSideCommandId,
+      uri => this.showNodeGraph(uri, true))
+    );
+    logger.debug("Registered " + PuppetNodeGraphToTheSideCommandId + " command");
+
+    this.provider = new NodeGraphContentProvider(connectionManager);
```

Comment:
Would it be better to pass in a `NodeGraphContentProvider` instead of an `IConnectionManager`, since the `ConnectionManager` object is only used by the `NodeGraphContentProvider`? This removes one reference, and allows us to change `ConnectionManager` independent of this class.

Or, does the array implementation of extensionFeatures in the main activation method not allow us to have dependencies between 'features'?

Reply:
Ideally I only wanted to pass in the LanguageClient object, however that needs a refactor of the connection manager which I don't want to do in this PR.

> Or, does the array implementation of extensionFeatures in the main activation method not allow us to have dependencies between 'features'?

Only becuase I've done it as a single line assignment.  There's nothing wrong with appending to the array on a per feature basis.